ios/engine/KMEI/KeymanEngine/Classes/Migrations.swift

```diff
@@ -236,60 +205,59 @@ public enum Migrations {
       let resources = possibleHistories[possibleHistories.count-1].resources
 
       resources.forEach { res in
-        if let kbd = res as? InstallableKeyboard {
-          var userKeyboards = Storage.active.userDefaults.userKeyboards
-
+        if let kbd = res as? FullKeyboardID {
           // Does not remove the deprecated keyboard's files - just the registration.
-          userKeyboards?.removeAll(where: { kbd2 in
-            return kbd.id == kbd2.id && kbd.languageID == kbd2.languageID
-          })
-          Storage.active.userDefaults.userKeyboards = userKeyboards
-        } else if let lex = res as? InstallableLexicalModel {
-          var userModels = Storage.active.userDefaults.userLexicalModels
-
+          if let match = userKeyboards.first(where: { $0.fullID == kbd }) {
+            if match.fullID == Defaults.keyboard.fullID {
+              currentDefaultKbdVersion = Version(match.version)
+            }
+            oldDefaultKbd = match
+          } else {
+            hasVersionDefaults = false
+          }
+        } else if let lex = res as? FullLexicalModelID {
           // Parallels the issue with deprecated files for keyboards.
-          userModels?.removeAll(where: { lex2 in
-            return lex.id == lex2.id && lex.languageID == lex2.languageID
-          })
-          Storage.active.userDefaults.userLexicalModels = userModels
+          if let match = userModels.first(where: { $0.fullID == lex }) {
+            if match.fullID == Defaults.lexicalModel.fullID {
+              currentDefaultLexVersion = Version(match.version)
+            }
+            oldDefaultLex = match
+          } else {
+            hasVersionDefaults = false
+          }
         } // else Not yet implemented
       }
+
+      // The user has customized their default resources; Keyman will refrain from
+      // changing the user's customizations.
+      if !hasVersionDefaults {
+        return
+      }
     }
 
     // Now to install the new version's resources.
-    var userKeyboards = Storage.active.userDefaults.userKeyboards ?? []
-    var userModels = Storage.active.userDefaults.userLexicalModels ?? []
     let defaultsNeedBackup = (lastVersion ?? Version.fallback) < Version.defaultsNeedBackup
-    var installKbd = false
-    var installLex = false
-
-    // Don't add the keyboard a second time if it's already installed and backed up.  Can happen
-    // if multiple Keyman versions match.
-    if !userKeyboards.contains(where: { kbd in
-      kbd.id == Defaults.keyboard.id && kbd.languageID == Defaults.keyboard.languageID
-    }) {
+
+    // First the keyboard.  If it needs an update:
+    if currentDefaultKbdVersion ?? Version("0.0")! <= Version(Defaults.keyboard.version)! || defaultsNeedBackup {
```

Comment:
What happens if `currentDefaultKbdVersion` is `nil`? That will end up installing the keyboard, correct?

Reply:
Exactly.  So, if a user is updating from KMW **and** still had their defaults installed, we'd reach this point.  KMW 10 had `european2` as a default keyboard, and for that case this _will_ be `nil` - there was no version of `sil_euro_latin` installed.  (At least, not by default.)

The intent is that `european2` is removed and replaced by `sil_euro_latin`, so we don't care what version of `european2` the user has installed - we're going to replace it.

Reply:
Gotcha.